File: LeetCode/46.php
<?php
class Solution
{
    function permute($nums): array
    {
        $count = count($nums);
        $used  = array_fill(0, $count, 0);

        return $this->dp($nums, $used, $count);;
    }

    public function dp(array $nums, array $used, int $count): array
    {
        $result = [];
        for ($i = 0; $i < $count; $i++) {
            if ($used[$i] != 1) {
                $used[$i] = 1;
                $temp = $this->dp($nums, $used, $count);
                $temp = array_map(function ($value) use ($nums, $i) {
                    array_unshift($value, $nums[$i]);
                    return $value;
                }, $temp);
                if ($temp == []) $temp[] = [$nums[$i]];
                $result = array_merge($result, $temp);
                $used[$i] = 0;
            }
        }
        return $result;
    }

    function permute1(array $nums): array
    {
        $result = [];
        $queue = new \SplQueue();
        for ($i = 0; $i < count($nums); $i++) {
            $this->swap($nums, $i, 0);
            $queue->enqueue($nums);
            $index = 0;
            while ($queue->count() != 0) {
                $count = $queue->count();
                $index += 1;
                while ($count--) {
                    $temp = $queue->bottom();
                    if ($index == count($nums))
                        $result[] = $temp;
                    $queue->dequeue();
                    for ($j = $index; $j < count($nums); $j++) {
                        $this->swap($temp, $j, $index);
                        $queue->enqueue($temp);
                        $this->swap($temp, $j, $index);
                    }
                }
            }
            $this->swap($nums, $i, 0);
        }
        return $result;
    }

    function permute2(array $nums): array
    {
        $result = $queue = [];
        $numsSize = count($nums);
        for ($i = 0; $i < $numsSize; $i++) {
            $this->swap($nums, 0, $i);
            $queue[] = $nums;
            $flag = 0;
            while (count($queue) != 0) {
                $flag++;
                if ($flag >= $numsSize - 1) {
                    $result = array_merge($result, $queue);
                    $queue = [];
                    break;
                }
                $queueSize = count($queue);
                while ($queueSize--) {
                    $temp = array_shift($queue);
                    for ($j = $flag; $j < $numsSize; $j++) {
                        $this->swap($temp, $flag, $j);
                        $queue[] = $temp;
                        $this->swap($temp, $flag, $j);
                    }
                }
            }
            $this->swap($nums, 0, $i);
        }
        return $result;
    }

    function swap(array &$nums, int $index1, int $index2)
    {
        $temp = $nums[$index1];
        $nums[$index1] = $nums[$index2];
        $nums[$index2] = $temp;
    }
}

$array = [1, 2, 3];
$test = new Solution();
var_dump($test->permute($array));

File: LeetCode/46.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class Solution
{
    public List<List<Integer>> permute(int[] nums) {
        List<Integer> numbers = Arrays.stream(nums).boxed().collect(Collectors.toList());
        List<List<Integer>> result = new ArrayList<>();
        List<List<Integer>> queue = new ArrayList<>();

        for (int i = 0; i < numbers.size(); i++) {
            swap(numbers, 0, i);
            queue.add(numbers);
            int flag = 0;

            while (queue.size() != 0) {
                flag++;

                if (flag == numbers.size()) {
                    result.addAll(queue);
                    queue.clear();
                    break;
                }
                int size = queue.size();

                while (size-- != 0) {
                    List<Integer> temp = queue.get(0);
                    queue.remove(0);

                    for (int j = flag; j < numbers.size(); j++) {
                        swap(temp, flag, j);
                        queue.add(Arrays.stream(temp.stream().mapToInt(Integer::valueOf).toArray())
                                .boxed()
                                .collect(Collectors.toList()));
                        swap(temp, flag, j);
                    }
                }
            }
            swap(numbers, 0, i);
        }
        return result;
    }

    public void swap(List<Integer> array, int idx1, int idx2) {
        int temp = array.get(idx1);
        array.set(idx1, array.get(idx2));
        array.set(idx2, temp);
    }

    public static void main(String[] args) {
        Solution test = new Solution();
        // test.permute(new int[] { 1, 2, 3 });
        System.out.println(test.permute(new int[] { 0, 1 }));
    }
}